File: LW1/python/LW1.py
import numpy as np
import matplotlib.pyplot as plt
import time


def generate_data(n=80, seed=42):
    """Генерация синтетических данных"""
    np.random.seed(seed)
    x = np.random.uniform(-3, 7, n)
    noise = np.random.normal(0, 0.3, n)
    y = 0.1 * (x - 4) * np.cos(x) + 0.5 * x + noise
    return x, y


def epanechnikov_kernel(z):
    """Ядро Епанечникова"""
    return np.where(np.abs(z) <= 1, 0.75 * (1 - z**2), 0.0)


def compute_delta(x):
    """Вычисление максимального расстояния между соседними точками (Δ)"""
    x_sorted = np.sort(x)
    deltas = np.diff(x_sorted)
    return np.max(deltas)


def nadaraya_watson_predict(x_train, y_train, x_pred, beta, delta):
    """Предсказание методом Надрая-Ватсона"""
    h = delta / beta
    y_pred = np.zeros_like(x_pred)
    
    for i, x0 in enumerate(x_pred):
        z = (x0 - x_train) / h
        K = epanechnikov_kernel(z)
        denominator = np.sum(K)
        
        if denominator > 1e-10:
            y_pred[i] = np.sum(K * y_train) / denominator
        else:
            # Если все веса нулевые — берём среднее
            y_pred[i] = np.mean(y_train)
    
    return y_pred


def loo_mse(x, y, beta, delta):
    """Leave-One-Out Cross Validation (скользящий экзамен)"""
    n = len(x)
    y_pred_loo = np.zeros(n)
    
    for i in range(n):
        x_train = np.delete(x, i)
        y_train = np.delete(y, i)
        y_pred_loo[i] = nadaraya_watson_predict(x_train, y_train, [x[i]], beta, delta)[0]
    
    mse = np.mean((y - y_pred_loo) ** 2)
    return mse


def find_best_beta(x, y, delta, beta_range=np.arange(0.1, 2.1, 0.1)):
    """Поиск оптимального beta через LOOCV"""
    mses = []
    for beta in beta_range:
        mse = loo_mse(x, y, beta, delta)
        mses.append(mse)
    
    best_idx = np.argmin(mses)
    best_beta = beta_range[best_idx]
    best_mse = mses[best_idx]
    
    return best_beta, best_mse, beta_range, mses


def main():
    print("=== Непараметрическая регрессия Nadaraya-Watson (LW1) ===\n")
    
    start_time = time.time()
    
    # 1. Генерация данных
    x, y = generate_data(n=80)
    delta = compute_delta(x)
    print(f"Максимальный шаг Δ = {delta:.4f}\n")
    
    # 2. Поиск лучшего beta
    print("Поиск оптимального параметра β...")
    best_beta, best_mse, betas, mses = find_best_beta(x, y, delta)
    
    print(f"Лучшее β = {best_beta:.1f}")
    print(f"Минимальная MSE (LOOCV) = {best_mse:.4f}\n")
    
    # 3. Построение финальной модели
    x_sorted = np.sort(x)
    y_pred = nadaraya_watson_predict(x, y, x_sorted, best_beta, delta)
    
    # 4. Визуализация
    plt.figure(figsize=(14, 10))
    
    # График 1: Данные + регрессия
    plt.subplot(2, 1, 1)
    plt.scatter(x, y, alpha=0.7, label='Исходные данные (с шумом)', color='blue')
    plt.plot(x_sorted, y_pred, 'r-', linewidth=2.5, label=f'НПР (β={best_beta:.1f})')
    plt.title('Непараметрическая регрессия Nadaraya-Watson')
    plt.xlabel('x')
    plt.ylabel('y')
    plt.legend()
    plt.grid(True, alpha=0.3)
    
    # График 2: Зависимость MSE от β
    plt.subplot(2, 1, 2)
    plt.plot(betas, mses, 'o-', color='green', linewidth=2)
    plt.axvline(x=best_beta, color='red', linestyle='--', alpha=0.7)
    plt.title('Зависимость ошибки LOOCV от параметра размытия β')
    plt.xlabel('β')
    plt.ylabel('MSE')
    plt.grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.show()
    
    end_time = time.time()
    print(f"Время выполнения: {end_time - start_time:.2f} секунд")


if __name__ == "__main__":
    main()    for i in range(n):
        mask = np.arange(n) != i
        y_pred_loo[i] = predict_nw(x[i], x[mask], y[mask], h)
    return np.mean((y - y_pred_loo)**2)

def find_best_beta(x, y, delta):
    """Перебор β ∈ [0.1, 2.0] шаг 0.1"""
    betas = np.arange(0.1, 2.01, 0.1)
    mses = [loo_mse(beta, x, y, delta) for beta in betas]
    best_idx = np.argmin(mses)
    return betas[best_idx], mses[best_idx], betas, mses

def main():
    start = time.time()
    
    # 1. Генерация данных
    x, y = generate_data()
    delta = compute_delta(x)
    print(f"Δ (max spacing) = {delta:.4f}")
    
    # 2. Поиск лучшего β через LOOCV
    best_beta, best_mse, betas, mses = find_best_beta(x, y, delta)
    print(f"Лучший β = {best_beta:.1f}, LOOCV MSE = {best_mse:.4f}")
    
    # 3. Финальная модель на всех данных
    h_opt = delta / best_beta
    x_plot = np.linspace(-3, 7, 500)
    y_plot = predict_nw(x_plot, x, y, h_opt)
    
    comp_time = time.time() - start
    print(f"Время вычислений: {comp_time:.3f} секунд")
    
    # 4. Графики
    fig, axs = plt.subplots(1, 2, figsize=(14, 6))
    
    axs[0].scatter(x, y, alpha=0.7, s=30, label='Данные (n=80)')
    axs[0].plot(x_plot, y_plot, 'r-', lw=2.5, label=f'НП регрессия (β={best_beta:.1f})')
    axs[0].set_xlabel('x')
    axs[0].set_ylabel('y')
    axs[0].set_title('Непараметрическая регрессия (Nadaraya-Watson)')
    axs[0].legend()
    axs[0].grid(True, alpha=0.3)
    
    axs[1].plot(betas, mses, 'bo-', markersize=4)
    axs[1].axvline(best_beta, color='red', linestyle='--', label=f'Лучший β = {best_beta:.1f}')
    axs[1].set_xlabel('Параметр размытия ядра β')
    axs[1].set_ylabel('MSE (leave-one-out)')
    axs[1].set_title('Зависимость ошибки от β')
    axs[1].legend()
    axs[1].grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.show()
    
    print(f"Оптимальная ширина окна h = {h_opt:.4f} (по формуле 4.7.5)")

if __name__ == "__main__":
    main()

File: LW1/cpp/LW1.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>
#include <chrono>
#include <fstream>
#include <cmath>
#include <iomanip>

using namespace std;

double kernel(double z) {
    return (fabs(z) <= 1.0) ? 0.75 * (1.0 - z * z) : 0.0;
}

double predict_nw(double x_new, const vector<double>& x_train, const vector<double>& y_train, double h) {
    double num = 0.0, den = 0.0;
    for (size_t i = 0; i < x_train.size(); ++i) {
        double z = (x_new - x_train[i]) / h;
        double k = kernel(z);
        num += k * y_train[i];
        den += k;
    }
    return (den > 1e-12) ? num / den : 0.0;
}

pair<vector<double>, vector<double>> generate_data(int n = 80) {
    mt19937 gen(42);
    uniform_real_distribution<double> ux(-3.0, 7.0);
    normal_distribution<double> noise(0.0, 0.3);
    vector<double> x(n), y(n);
    for (int i = 0; i < n; ++i) {
        x[i] = ux(gen);
        y[i] = 0.1 * (x[i] - 4.0) * cos(x[i]) + 0.5 * x[i] + noise(gen);
    }
    return {x, y};
}

double compute_delta(const vector<double>& x) {
    vector<double> xs = x;
    sort(xs.begin(), xs.end());
    double max_d = 0.0;
    for (size_t i = 1; i < xs.size(); ++i) {
        max_d = max(max_d, xs[i] - xs[i - 1]);
    }
    return max_d;
}

double loo_mse(double beta, const vector<double>& x, const vector<double>& y, double delta) {
    size_t n = x.size();
    double h = delta / beta;
    double mse = 0.0;
    for (size_t i = 0; i < n; ++i) {
        vector<double> xt, yt;
        for (size_t j = 0; j < n; ++j) {
            if (j != i) {
                xt.push_back(x[j]);
                yt.push_back(y[j]);
            }
        }
        double pred = predict_nw(x[i], xt, yt, h);
        mse += (y[i] - pred) * (y[i] - pred);
    }
    return mse / n;
}

int main() {
    std::setlocale(LC_ALL, "ru_RU.UTF-8");
    std::cout << std::fixed << std::setprecision(4);
    auto start = chrono::high_resolution_clock::now();

    cout << "=== Непараметрическая регрессия (C++ / Nadaraya-Watson) ===\n\n";

    auto [x, y] = generate_data();
    double delta = compute_delta(x);
    cout << "Δ (max spacing) = " << fixed << setprecision(4) << delta << "\n";

    // Перебор β
    vector<double> betas, mses;
    double best_beta = 0.1, best_mse = 1e9;
    for (double b = 0.1; b <= 2.01; b += 0.1) {
        double mse = loo_mse(b, x, y, delta);
        betas.push_back(b);
        mses.push_back(mse);
        if (mse < best_mse) {
            best_mse = mse;
            best_beta = b;
        }
    }

    cout << "Лучший β = " << best_beta << "\n";
    cout << "LOOCV MSE = " << fixed << setprecision(4) << best_mse << "\n\n";

    // Финальная модель
    double h_opt = delta / best_beta;
    vector<double> x_plot, y_plot;
    for (double xi = -3.0; xi <= 7.0; xi += 0.02) {
        x_plot.push_back(xi);
        y_plot.push_back(predict_nw(xi, x, y, h_opt));
    }

    // Сохранение для графиков
    ofstream fdata("data_points.csv");
    for (size_t i = 0; i < x.size(); ++i) fdata << x[i] << "," << y[i] << "\n";
    fdata.close();

    ofstream freg("regression_curve.csv");
    for (size_t i = 0; i < x_plot.size(); ++i) freg << x_plot[i] << "," << y_plot[i] << "\n";
    freg.close();

    ofstream fmse("mse_vs_beta.csv");
    for (size_t i = 0; i < betas.size(); ++i) fmse << betas[i] << "," << mses[i] << "\n";
    fmse.close();

    auto end = chrono::high_resolution_clock::now();
    double sec = chrono::duration<double>(end - start).count();
    cout << "Время выполнения: " << fixed << setprecision(2) << sec << " сек\n";
    cout << "Данные сохранены в *.csv файлы\n";

    return 0;
}
